python/cudf/cudf/_lib/parquet.pyx

```diff
@@ -361,9 +361,12 @@ def write_parquet(
 
     for i, name in enumerate(table._column_names, num_index_cols_meta):
         if not isinstance(name, str):
-            raise ValueError("parquet must have string column names")
-
-        tbl_meta.get().column_metadata[i].set_name(name.encode())
+            if cudf.get_option("mode.pandas_compatible"):
+                tbl_meta.get().column_metadata[i].set_name(str(name).encode())
+            else:
+                raise ValueError("parquet must have string column names")
```

Comment:
I think we should have a very strong reason to differ from pandas behavior here. Do we raise an error here as a form of "user protection" to prevent unwanted outcomes from non-string column names? I would favor matching pandas and always casting to a string if that's what is expected by pandas.

Reply:
Or is this behavior that should be gated by a pandas version check?

Reply:
@vuule and rest of the IO team thought it'd be best to not allow this behavior and have this behavior of converting the column names to string only in pandas compatibility mode. I agree we need to generally match pandas by default but this can become quite erroneous for us very quickly because cases when trying to write this type of dataframe even pandas exhibit un-defined behavior:
```python
In [10]: df = cudf.DataFrame({1: [1, 2, 3], "1": ["a", "b", "c"]})

In [11]: df
Out[11]: 
   1  1
0  1  a
1  2  b
2  3  c
```

Reply:
> Or is this behavior that should be gated by a pandas version check?

Not quite in favor of this because it's not quite a compelling one that we have to behave differently because of a different pandas version installed.

Reply:
> Do we raise an error here as a form of "user protection" to prevent unwanted outcomes from non-string column names?

Yes. As @galipremsagar mentioned, casting all column names to strings creates an opportunity for errors. Alternatively, we could check if casting is safe, but this check would not be trivially cheap.
We are far from 100% Pandas compatible in cuIO; there are options that would completely kill performance and some that don't really translate to GPU. The fact that there's an option to be compatible here sounds great to me.

I'll gladly discuss this further on Monday.

Reply:
Got it. This is well justified, I just didn’t know that context. We should make sure these kinds of differences are documented clearly. It relates loosely to the section on column names, so maybe describe it there or in the API docs for affected functions if it’s not already mentioned: https://docs.rapids.ai/api/cudf/stable/user_guide/pandas-comparison/#column-names

Reply:
Thanks @bdice & @vuule ! I have documented this difference in behavior and how to work-around it here: https://github.com/rapidsai/cudf/pull/13505/commits/d9fc2f51b4123e068af4391b0659d8655d686b58